File: 21. Merge Two Sorted Lists/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    var head1 = new ListNode(0)
    var cur1 = l1
    var cur2 = l2
    var pre1 = head1
    pre1.next = cur1
    while(cur1 && cur2) {
        if(cur2.val > cur1.val) {
            pre1 = cur1
            cur1 = cur1.next
        } else {
            let temp = cur2.next
            pre1.next = cur2
            cur2.next = cur1
            pre1 = cur2
            cur2 = temp
        }
    }
    if(cur2) {
        pre1.next = cur2
    }
    return head1.next
};

File: 21. Merge Two Sorted Lists/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
    if l1 == nil {
        return l2
    }
    if l2 ==nil {
        return l1
    }
    var headNode ListNode
    headNode.Val = 0
    headNode.Next = nil
    var head *ListNode = &headNode
    var cur *ListNode
    if l1.Val >= l2.Val {
        head.Next = l2
        l2 = l2.Next
    } else {
        head.Next = l1
        l1 = l1.Next
    }
    cur = head.Next
    for {
        if l1 == nil {
            cur.Next = l2
            break
        }
        if l2 == nil {
            cur.Next = l1
            break
        }
        if l1.Val >= l2.Val {
            cur.Next = l2
            cur = l2
            l2 = l2.Next
        } else {
            cur.Next = l1
            cur = l1
            l1 = l1.Next
        }
    }
    return head.Next
}